# Vehicle Count Prediction From Sensor Data

Sensors at road junctions collect vehicle count data at different times which helps transport managers make informed decisions. Here we will predict vehicle count based on this sensor data using machine learning techniques.

## Implementation of Vehicle Count Prediction

Datetime: Timestamp of when data was collected.

Vehicles: Number of vehicles at that timestamp.

Since Vehicles is a numeric attribute, using Regression would be the best choice. Regression is a supervised learning technique used to predict a numeric value based on historical data. So we will use Random Forest Regressor that predicts the target by averaging the results of multiple decision trees.

### Step 1: Import Libraries and Load Data

We will be using pandas library for data manipulation and loading dataset into a DataFrame.

In [1]:
import pandas as pd
train = pd.read_csv('vehicles.csv')
train.head()

DateTime  Vehicles
0  2015-11-01 00:00:00        15
1  2015-11-01 01:00:00        13
2  2015-11-01 02:00:00        10
3  2015-11-01 03:00:00         7
4  2015-11-01 04:00:00         9

### Step 2: Feature Extraction from Datetime

We extract useful features from Datetime attribute such as day, weekday, hour, month, year etc which will help in predicting vehicle count.

get_dom(dt): Extracts day of the month from the timestamp dt (e.g. for "2021-01-15", it would return 15).

get_weekday(dt): Extracts weekday from the timestamp where Monday is 0, Tuesday is 1 and so on (e.g. "2021-01-15" would return 4 for Friday).

get_hour(dt): Extracts hour from the timestamp dt (e.g. "2021-01-15 15:30:00" would return 15).

get_year(dt): Extracts year from the timestamp dt (e.g. "2021-01-15" would return 2021).

train['DateTime'].map(pd.to_datetime): Converts DateTime column into a proper datetime format so we can use date-time attributes.

train['date'] = train['DateTime'].map(get_dom): Creates a new column date that holds day of the month extracted from the DateTime column.

train['weekday'] = train['DateTime'].map(get_weekday): Creates a new column weekday that holds weekday number like Monday=0.

train['hour'] = train['DateTime'].map(get_hour): Creates a new column hour that holds hour of the timestamp in 24-hour format.

In [2]:
def get_dom(dt):
	return dt.day
def get_weekday(dt):
	return dt.weekday()
def get_hour(dt):
	return dt.hour
def get_year(dt):
	return dt.year
def get_month(dt):
	return dt.month
def get_dayofyear(dt):
	return dt.dayofyear
def get_weekofyear(dt):
	return dt.weekofyear

train['DateTime'] = train['DateTime'].map(pd.to_datetime)
train['date'] = train['DateTime'].map(get_dom)
train['weekday'] = train['DateTime'].map(get_weekday)
train['hour'] = train['DateTime'].map(get_hour)
train['month'] = train['DateTime'].map(get_month)
train['year'] = train['DateTime'].map(get_year)
train['dayofyear'] = train['DateTime'].map(get_dayofyear)
train['weekofyear'] = train['DateTime'].map(get_weekofyear)

train.head()

DateTime  Vehicles  date  weekday  hour  month  year  dayofyear  \
0 2015-11-01 00:00:00        15     1        6     0     11  2015        305   
1 2015-11-01 01:00:00        13     1        6     1     11  2015        305   
2 2015-11-01 02:00:00        10     1        6     2     11  2015        305   
3 2015-11-01 03:00:00         7     1        6     3     11  2015        305   
4 2015-11-01 04:00:00         9     1        6     4     11  2015        305   

   weekofyear  
0          44  
1          44  
2          44  
3          44  
4          44

### Step 3: Separate Class Label (Vehicles)

We separate Vehicles class label and store it in the target variable while keeping remaining features in train1 variable.

In [3]:
train = train.drop(['DateTime'], axis=1)
train1 = train.drop(['Vehicles'], axis=1)
target = train['Vehicles']
print(train1.head())
target.head()

   date  weekday  hour  month  year  dayofyear  weekofyear
0     1        6     0     11  2015        305          44
1     1        6     1     11  2015        305          44
2     1        6     2     11  2015        305          44
3     1        6     3     11  2015        305          44
4     1        6     4     11  2015        305          44


0    15
1    13
2    10
3     7
4     9
Name: Vehicles, dtype: int64

Result shows data where various time-based features like day of the month, weekday, hour, month, year, day of the year and week of the year are extracted from the DateTime column. Vehicles column shows vehicle count at that timestamp which is 15.

### Step 4: Train the Model Using Random Forest Regressor

We will train RandomForestRegressor model using features and target variable.

In [5]:
from sklearn.ensemble import RandomForestRegressor
m1=RandomForestRegressor()
m1.fit(train1,target)
m1.predict([[11,6,0,1,2015,11,2]])

d:\mini projects\projects\Lib\site-packages\sklearn\utils\validation.py:2691: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


array([9.61021429])

It show that based on input features model predicts approximately 9.61 vehicles at that specific timestamp.